src/job_queue/mod.rs

```diff
@@ -23,8 +24,354 @@
+//!
+//!     let job = HelloJobAsync("world".to_string());
+//!     let job_handle = job_queue.add_job_mut(job, 1)?;
+//!
+//!     // await job to complete and obtain the (wrapped) job result
+//!     let completion = job_handle.await?;
+//!     let output = JobResultWrapper::<String>::try_from(completion)?.into_inner();
+//!
+//!     assert_eq!("hello world", output);
+//!
+//!     Ok(())
+//! }
+//! ```
+//!
+//! ## async vs blocking Job.
+//!
+//! To demonstrate the difference, we will implement an example job first as an
+//! async job and then as a blocking job.
+//!
+//! The example job finds all the prime numbers in a provided range.
+//!
+//! We create 100 jobs, each searching a range of 100 numbers. So the first
+//! 10000 integers are searched by all jobs.
+//!
+//! We use a `JobQueue<QueueJobPriority>` where `QueueJobPriority` is an enum we
+//! define that simply has `Low` and `High` variants.
+//!
+//! The jobs are added to the queue in ascending order but each is assigned a
+//! random job priority (either High or Low).  High priority jobs will process
+//! first, thus queue-processing order will not match the order of adding jobs.
+//!
+//! In this example, job results are obtained by awaiting each JobHandle in the
+//! order it was added.  Thus results are obtained in FIFO order despite the
+//! out-of-order processing.
+//!
+//! Alternatively a JoinSet or join_all() could be used to await all job-handles
+//! simultaneously and obtain results in queue-processing order as they complete.
+//!
+//! Likewise in an application with many concurrent tasks, each task might be
+//! submitting a job and immediately awaiting the JobHandle.  In that scenario
+//! whichever task has submitted the highest priority job will obtain results
+//! first.
+//!
+//! ### Async Job considerations
+//!
+//! 1. the Job::is_async() impl returns true.
+//! 2. Job::run_async() or Job::run_async_cancellable() must be implemented.
+//!
+//! It is important the job yield regularly to the async runtime.  Our
+//! processing is inherently blocking, so we accomplish this simply by making
+//! the is_prime() fn async, which is called in every loop iteration.
+//!
+//! ### Blocking job considerations
+//!
+//! 1. the Job::is_async() impl returns false.
+//! 2. Job::run() must be implemented.
+//! 3. it is necessary to regularly poll for a job-cancellation message in the
+//!    job's main processing loop.
+//!
+//! ### Example
+//!
+//! ```
+//! use neptune_cash::job_queue::JobCompletion;
+//! use neptune_cash::job_queue::JobResultWrapper;
+//! use neptune_cash::job_queue::JobQueue;
+//! use neptune_cash::job_queue::channels::JobCancelReceiver;
+//! use neptune_cash::job_queue::traits::*;
+//! use rand::Rng;
+//!
+//! // ### First lets define some common types ###
+//! // -------------------------------------------
+//!
+//! // define job priority levels for this job-queue.
+//! // note that the queue could process only one type of
```

Comment:
Maybe “is able to” instead of “could”?

Reply:
addressed in 8b189964d2c0fd51b8e9c4f7899b2f713bc95b99.  removed superfluous line.